libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+
+bool setup_debug_privileges(bool b) {
+	HANDLE tok;
+	if (!OpenProcessToken (GetCurrentProcess (), TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY, &tok)) {
+		return false;
+	}
+	bool ret = false;
+	LUID luid;
+	if (LookupPrivilegeValue (NULL, SE_DEBUG_NAME, &luid)) {
+		TOKEN_PRIVILEGES tp;
+		tp.PrivilegeCount = 1;
+		tp.Privileges[0].Luid = luid;
+		tp.Privileges[0].Attributes = b ? SE_PRIVILEGE_ENABLED : 0;
+		if (AdjustTokenPrivileges (tok, FALSE, &tp, 0, NULL, NULL)) {
+			// TODO: handle ERROR_NOT_ALL_ASSIGNED
+			ret = GetLastError () == ERROR_SUCCESS;
+		}
+	}
+	CloseHandle (tok);
+	return ret;
+}
+
+int w32_init(RDebug *dbg) {
+	RIOW32Dbg *rio = dbg->user = R_NEW (RIOW32Dbg);
+	if (!rio) {
+		eprintf ("w32_init: failed to allocate memory\n");
+		return false;
+	}
+	setup_debug_privileges (true);
+	// rio->dbgpriv = setup_debug_privileges (true);
+	rio->ph = (HANDLE)NULL;
+	rio->debug = false;
+
+	HMODULE lib = GetModuleHandle ("kernel32"); //Always loaded
+	w32_GetEnabledXStateFeatures = GetProcAddress (lib, "GetEnabledXStateFeatures");
+	w32_InitializeContext = GetProcAddress (lib, "InitializeContext");
+	w32_GetXStateFeaturesMask = GetProcAddress (lib, "GetXStateFeaturesMask");
+	w32_LocateXStateFeature = GetProcAddress (lib, "LocateXStateFeature");
+	w32_SetXStateFeaturesMask = GetProcAddress (lib, "SetXStateFeaturesMask");
+
+	lib = GetModuleHandle ("ntdll"); //Always loaded
+	w32_NtQueryInformationThread = GetProcAddress (lib, "NtQueryInformationThread");
+	return true;
+}
+
+static int suspend_thread(HANDLE th, int bits) {
+	int ret;
+	//if (bits == R_SYS_BITS_32) {
+		if ((ret = SuspendThread (th)) == -1) {
+			r_sys_perror ("suspend_thread/SuspendThread");
+		}
+	/*} else {
+		if ((ret = Wow64SuspendThread (th)) == -1) {
+			r_sys_perror ("suspend_thread/Wow64SuspendThread");
+		}
+	}*/
+	return ret;
+}
+
+static int resume_thread(HANDLE th, int bits) {
+	int ret;
+	//if (bits == R_SYS_BITS_32) {
+		if ((ret = ResumeThread (th)) == -1) {
+			r_sys_perror ("resume_thread/ResumeThread");
+		}
+	/*} else {
+		if ((ret = ResumeThread (th)) == -1) {
+			r_sys_perror ("resume_thread/Wow64ResumeThread");
+		}
+	}*/
+	return ret;
+}
+
+static int set_thread_context(HANDLE th, const ut8 *buf, int size, int bits) {
+	bool ret;
+	//if (bits == R_SYS_BITS_32) {
+		CONTEXT ctx = {0};
+		if (size > sizeof (ctx)) {
+			size = sizeof (ctx);
+		}
+		memcpy (&ctx, buf, size);
+		if(!(ret = SetThreadContext (th, &ctx))) {
+			r_sys_perror ("set_thread_context/SetThreadContext");
+		}
+	/*} else {
+		WOW64_CONTEXT ctx = {0};
+		if (size > sizeof (ctx)) {
+			size = sizeof (ctx);
+		}
+		memcpy (&ctx, buf, size);
+		if(!(ret = Wow64SetThreadContext (th, &ctx))) {
+			r_sys_perror ("set_thread_context/Wow64SetThreadContext");
+		}
+	}*/
+	return ret;
+}
+
+static int get_thread_context(HANDLE th, ut8 *buf, int size, int bits) {
+	int ret = 0;
+	//if (bits == R_SYS_BITS_32) {
```

Comment:
i dont like all this commented code. looks to me like debugging 32bit processes on 64bit builds is no longer working

Reply:
I wasn't entirely sure if its needed or not because documentation isn't clear so I've commented it out in the hopes that maybe someone who knows better than me will sort it out.